### DISCUSSION TITLE
Optional file uploads

### DISCUSSION THREAD

**Discussion by daleif** [mojolicious/mojo#1972] [Ideas]
Since a file upload field in a `multipart/form-data` form always generates an `Mojo::Upload` object even if there is no file specified, then

```
$v->optional('my_file','not_empty');
```

will never remove `my_file` if there was no file specified in the upload. 

I'm wondering whether it would make sense to change the `not_empty` filter to something similar to 

```
not_empty => sub ($v, $name, $value) {
	    if ( ref $value && $value->can('size') ) {
		$value->size ? $value : ()
	    } else {
		length $value ? $value : ()
	    }
	}
```

Then at least it also works with objects that has a `->size` method like `Mojo::Upload`.

`not_empty` is currently implemented using the `length` line.

**Comment by kraih** [comment]
Would be consistent with `$v->size` i guess if it worked for `Mojo::Upload` objects.
